lib/gears/reactive.lua

```diff
@@ -0,0 +1,647 @@
+--   instances *should* be fine.
+-- * This module will **NOT** try to track the change of other
+--   functions and methods called by the expression. It is **NOT** recursive
+--   and only the top level properties are tracked. This is a feature, not a
+--   bug. If it was recusrsive, this list of limitations or gotchas would be
+--   endless.
+-- * There is currently no way to disable a reactive expression once it's
+--   been defined. This may change eventually.
+-- * Rio Lua 5.1 (not LuaJIT 2.1) is currently not support. If you need it,
+--   report a bug.
+--
+-- @author Emmanuel Lepage-Vallee &lt;elv1313@gmail.com&gt;
+-- @copyright 2017-2020 Emmanuel Lepage-Vallee
+-- @classmod gears.reactive
+
+-- This file is provided under the BSD 2 clause license since it is better than
+-- any existing implementation (most of which only support Lua 5.1).
+--
+-- Copyright 2020 Emmanuel Lepage-Vallee &lt;elv1313@gmail.com&gt;
+--
+-- Redistribution and use in source and binary forms, with or without modification,
+-- are permitted provided that the following conditions are met:
+--
+-- 1. Redistributions of source code must retain the above copyright notice, this
+--    list of conditions and the following disclaimer.
+--
+-- 2. Redistributions in binary form must reproduce the above copyright notice,
+--    this list of conditions and the following disclaimer in the documentation
+--    and/or other materials provided with the distribution.
+--
+-- THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+-- AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+-- IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+-- DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+-- FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+-- (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+-- OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+-- THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+-- NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+-- IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+local gobject = require("gears.object")
+local gtable = require("gears.table")
+local gtimer = require("gears.timer")
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+-- It's only possible to know when something changes if it has signals.
+local function is_gobject(input)
+    return type(input) == "table" and rawget(input, "emit_signal")
+end
+
+-- If the sandbox proxies were to leave the reactive expressions, bad
+-- things would happen. Catch functions and constructors to make sure
+-- we strip away the proxy.
+local function is_callable(object)
+    local t = type(object)
+
+    if t == "function" then return true  end
+    if t ~= "table"    then return false end
```

Comment:
What about userdata? Those can also be callable through a metatable.

(I can also make booleans callable with a metatable, but let's ignore that...)

Reply:
> What about userdata?

"maybe fixed". I admit I didn't test this much beyond basic client properties. `gears.reactive` is pretty hard to use with native capi objects... It "works", but there is no reliable way to detect changes to let say, `client.focus` or `awful.screen.focused()` using the "standard" property magic. You have to use the "hub object" concept to mitigate that (as mentioned in the doc). If you do that, you might as well put the values directly in the hub object rather than the client object?

> (I can also make booleans callable with a metatable, but let's ignore that...)

And yet I was crazy enough to do that when replacing some getter/setter methods with properties... Oh well, in retrospect that wasn't the best idea ever...

Reply:
> And yet I was crazy enough to do that when replacing some getter/setter methods with properties...

Wait, what? Where? You cannot "specialise" booleans. Setting the metatable of one boolean sets the metatable of all 
(current and future) booleans.